File: python/classes/vending_machine.py
"""Vending Machine — a finite state machine with coin-constrained change-making.

First project in the Rosetta Stone where operations are gated by the class's
STATE, not (only) by its DATA. `insert_coin`, `select`, and `refund` are legal
in different states and illegal in others. Purchase is an atomic transition
that touches three pieces of state simultaneously: slot inventory, machine
coin inventory (consuming change coins, adding inserted coins), and the
state itself.

Change-making reuses G007's greedy algorithm but is constrained by the
machine's current coin stock: if it can't assemble the owed amount from
what it has, the purchase fails cleanly (coins refunded, state resets).
"""

from __future__ import annotations
from dataclasses import dataclass, field
from enum import Enum


class MachineState(Enum):
    IDLE = "idle"              # no coins inserted
    ACCEPTING = "accepting"    # coins inserted, waiting for selection


# Standard US coin denominations (cents). Ordered high → low for greedy change.
COINS_CENTS = (100, 25, 10, 5, 1)


@dataclass
class Slot:
    slot_id: str
    product: str
    price_cents: int
    quantity: int


class VendingError(Exception): ...
class UnknownSlot(VendingError): ...
class SoldOut(VendingError): ...
class InsufficientFunds(VendingError): ...
class CannotMakeChange(VendingError): ...
class BadState(VendingError): ...
class InvalidCoin(VendingError): ...


@dataclass
class VendingMachine:
    slots: dict[str, Slot] = field(default_factory=dict)
    coin_inventory: dict[int, int] = field(
        default_factory=lambda: {d: 0 for d in COINS_CENTS})
    inserted_cents: int = 0
    inserted_coins: dict[int, int] = field(
        default_factory=lambda: {d: 0 for d in COINS_CENTS})
    state: MachineState = MachineState.IDLE

    # --- setup ---
    def add_slot(self, slot: Slot) -> None:
        if slot.slot_id in self.slots:
            raise ValueError(f"slot already exists: {slot.slot_id}")
        if slot.price_cents <= 0:
            raise ValueError("price must be positive")
        self.slots[slot.slot_id] = slot

    def load_change(self, denom: int, count: int) -> None:
        if denom not in COINS_CENTS:
            raise InvalidCoin(f"not a recognized denomination: {denom}")
        if count < 0:
            raise ValueError("count must be non-negative")
        self.coin_inventory[denom] = self.coin_inventory.get(denom, 0) + count

    # --- interactive operations ---
    def insert_coin(self, denom: int) -> int:
        """Accept a coin. Transitions IDLE→ACCEPTING on the first coin.
        Returns new inserted total."""
        if self.state not in (MachineState.IDLE, MachineState.ACCEPTING):
            raise BadState(f"cannot insert coin while {self.state.value}")
        if denom not in COINS_CENTS:
            raise InvalidCoin(f"not a recognized denomination: {denom}")
        self.inserted_cents += denom
        self.inserted_coins[denom] = self.inserted_coins.get(denom, 0) + 1
        self.state = MachineState.ACCEPTING
        return self.inserted_cents

    def refund(self) -> dict[int, int]:
        """Return all inserted coins. Transitions back to IDLE. Legal in any state."""
        refunded = dict(self.inserted_coins)
        self.inserted_cents = 0
        self.inserted_coins = {d: 0 for d in COINS_CENTS}
        self.state = MachineState.IDLE
        return {d: c for d, c in refunded.items() if c > 0}

    def select(self, slot_id: str) -> tuple[str, dict[int, int]]:
        """Atomically dispense product + change. Returns (product, change_coins).

        Transitions: ACCEPTING → IDLE on success. On failure, state is unchanged
        and coins remain inserted (caller may refund or insert more).
        """
        if self.state != MachineState.ACCEPTING:
            raise BadState(f"cannot select while {self.state.value}")
        if slot_id not in self.slots:
            raise UnknownSlot(f"unknown slot: {slot_id}")
        slot = self.slots[slot_id]
        if slot.quantity <= 0:
            raise SoldOut(f"{slot_id} ({slot.product}) is sold out")
        if self.inserted_cents < slot.price_cents:
            raise InsufficientFunds(
                f"{slot.product} is {slot.price_cents}¢, "
                f"{self.inserted_cents}¢ inserted")

        change_owed = self.inserted_cents - slot.price_cents
        # Merge inserted coins into the machine's available change pool
        # SO the machine can potentially use them to make change.
        pool = dict(self.coin_inventory)
        for d, c in self.inserted_coins.items():
            pool[d] = pool.get(d, 0) + c
        change = _greedy_change(change_owed, pool)
        if change is None:
            raise CannotMakeChange(
                f"cannot make {change_owed}¢ from available coins; refunding")

        # Atomic commit: subtract change from pool, update inventory accordingly.
        for d, c in change.items():
            pool[d] -= c
        self.coin_inventory = pool
        slot.quantity -= 1
        self.inserted_cents = 0
        self.inserted_coins = {d: 0 for d in COINS_CENTS}
        self.state = MachineState.IDLE
        return (slot.product, {d: c for d, c in change.items() if c > 0})

    # --- queries ---
    def available(self, slot_id: str) -> bool:
        return slot_id in self.slots and self.slots[slot_id].quantity > 0

    def total_coins_cents(self) -> int:
        return sum(d * c for d, c in self.coin_inventory.items())


def _greedy_change(amount: int, pool: dict[int, int]) -> dict[int, int] | None:
    """Try to make `amount` cents from `pool` using high-denom-first greedy.
    Returns the coin counts used, or None if impossible with these coins."""
    result: dict[int, int] = {}
    remaining = amount
    for d in sorted(pool.keys(), reverse=True):
        if d > remaining:
            continue
        use = min(pool[d], remaining // d)
        if use > 0:
            result[d] = use
            remaining -= d * use
        if remaining == 0:
            break
    return result if remaining == 0 else None


# --- demo ---
if __name__ == "__main__":
    m = VendingMachine()
    m.add_slot(Slot("A1", "Cola",    price_cents=125, quantity=5))
    m.add_slot(Slot("B2", "Chips",   price_cents=100, quantity=3))
    m.add_slot(Slot("C3", "Gum",     price_cents=50,  quantity=10))
    m.load_change(25, 20)
    m.load_change(10, 20)
    m.load_change(5, 20)
    m.load_change(1, 50)

    # Try to select without coins — state gate refuses.
    try:
        m.select("A1")
    except BadState as e:
        print(f"refused: {e}")

    # Insert $1.50, buy Cola ($1.25), get 25¢ change.
    m.insert_coin(100)
    m.insert_coin(25)
    m.insert_coin(25)
    print(f"inserted: {m.inserted_cents}¢  state: {m.state.value}")
    product, change = m.select("A1")
    print(f"dispensed: {product}; change: {change}  state: {m.state.value}")

    # Deliberately run down change and try a purchase the machine can't fulfill.
    m2 = VendingMachine()
    m2.add_slot(Slot("G1", "Gum", price_cents=30, quantity=10))
    m2.load_change(25, 0)
    m2.load_change(10, 0)
    m2.load_change(5, 0)          # no small change at all
    m2.load_change(1, 0)
    m2.insert_coin(100)
    try:
        m2.select("G1")
    except CannotMakeChange as e:
        print(f"refused: {e}")
        refunded = m2.refund()
        print(f"refunded: {refunded}")

File: rust/classes/vending_machine.rs
//! Vending Machine — a finite state machine with coin-constrained change-making.

use std::collections::HashMap;

pub const COINS_CENTS: &[i64] = &[100, 25, 10, 5, 1];

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MachineState {
    Idle,
    Accepting,
}

#[derive(Debug, Clone)]
pub struct Slot {
    pub slot_id: String,
    pub product: String,
    pub price_cents: i64,
    pub quantity: i64,
}

#[derive(Debug)]
pub enum VendingError {
    UnknownSlot(String),
    DuplicateSlot(String),
    InvalidPrice,
    SoldOut(String),
    InsufficientFunds { owed: i64, inserted: i64 },
    CannotMakeChange(i64),
    BadState(String),
    InvalidCoin(i64),
    InvalidCount,
}

pub struct VendingMachine {
    pub slots: HashMap<String, Slot>,
    pub coin_inventory: HashMap<i64, i64>,
    pub inserted_cents: i64,
    pub inserted_coins: HashMap<i64, i64>,
    pub state: MachineState,
}

impl VendingMachine {
    pub fn new() -> Self {
        let coins = COINS_CENTS.iter().map(|&d| (d, 0)).collect();
        let inserted = COINS_CENTS.iter().map(|&d| (d, 0)).collect();
        Self {
            slots: HashMap::new(),
            coin_inventory: coins,
            inserted_cents: 0,
            inserted_coins: inserted,
            state: MachineState::Idle,
        }
    }

    pub fn add_slot(&mut self, slot: Slot) -> Result<(), VendingError> {
        if self.slots.contains_key(&slot.slot_id) {
            return Err(VendingError::DuplicateSlot(slot.slot_id));
        }
        if slot.price_cents <= 0 {
            return Err(VendingError::InvalidPrice);
        }
        self.slots.insert(slot.slot_id.clone(), slot);
        Ok(())
    }

    pub fn load_change(&mut self, denom: i64, count: i64) -> Result<(), VendingError> {
        if !COINS_CENTS.contains(&denom) { return Err(VendingError::InvalidCoin(denom)); }
        if count < 0 { return Err(VendingError::InvalidCount); }
        *self.coin_inventory.entry(denom).or_insert(0) += count;
        Ok(())
    }

    pub fn insert_coin(&mut self, denom: i64) -> Result<i64, VendingError> {
        match self.state {
            MachineState::Idle | MachineState::Accepting => {}
        }
        if !COINS_CENTS.contains(&denom) {
            return Err(VendingError::InvalidCoin(denom));
        }
        self.inserted_cents += denom;
        *self.inserted_coins.entry(denom).or_insert(0) += 1;
        self.state = MachineState::Accepting;
        Ok(self.inserted_cents)
    }

    pub fn refund(&mut self) -> HashMap<i64, i64> {
        let refund: HashMap<i64, i64> = self
            .inserted_coins
            .iter()
            .filter(|&(_, &c)| c > 0)
            .map(|(&d, &c)| (d, c))
            .collect();
        self.inserted_cents = 0;
        for (_, c) in self.inserted_coins.iter_mut() { *c = 0; }
        self.state = MachineState::Idle;
        refund
    }

    pub fn select(&mut self, slot_id: &str) -> Result<(String, HashMap<i64, i64>), VendingError> {
        if self.state != MachineState::Accepting {
            return Err(VendingError::BadState(format!("cannot select while {:?}", self.state)));
        }
        let (product, price_cents) = {
            let slot = self.slots.get(slot_id)
                .ok_or_else(|| VendingError::UnknownSlot(slot_id.to_string()))?;
            if slot.quantity <= 0 {
                return Err(VendingError::SoldOut(slot_id.to_string()));
            }
            if self.inserted_cents < slot.price_cents {
                return Err(VendingError::InsufficientFunds {
                    owed: slot.price_cents, inserted: self.inserted_cents,
                });
            }
            (slot.product.clone(), slot.price_cents)
        };
        let change_owed = self.inserted_cents - price_cents;

        // Pool = machine coins + inserted coins (inserted coins are now the machine's).
        let mut pool: HashMap<i64, i64> = self.coin_inventory.clone();
        for (&d, &c) in &self.inserted_coins {
            *pool.entry(d).or_insert(0) += c;
        }
        let change = greedy_change(change_owed, &pool)
            .ok_or(VendingError::CannotMakeChange(change_owed))?;

        // Commit: subtract change from pool, update inventory, dispense product, reset.
        for (&d, &c) in &change {
            *pool.get_mut(&d).unwrap() -= c;
        }
        self.coin_inventory = pool;
        self.slots.get_mut(slot_id).unwrap().quantity -= 1;
        self.inserted_cents = 0;
        for (_, c) in self.inserted_coins.iter_mut() { *c = 0; }
        self.state = MachineState::Idle;
        Ok((product, change.into_iter().filter(|&(_, c)| c > 0).collect()))
    }
}

impl Default for VendingMachine {
    fn default() -> Self { Self::new() }
}

fn greedy_change(amount: i64, pool: &HashMap<i64, i64>) -> Option<HashMap<i64, i64>> {
    let mut result = HashMap::new();
    let mut remaining = amount;
    let mut denoms: Vec<i64> = pool.keys().cloned().collect();
    denoms.sort_by(|a, b| b.cmp(a));
    for d in denoms {
        if d > remaining { continue; }
        let available = *pool.get(&d).unwrap_or(&0);
        let use_count = available.min(remaining / d);
        if use_count > 0 {
            result.insert(d, use_count);
            remaining -= d * use_count;
        }
        if remaining == 0 { break; }
    }
    if remaining == 0 { Some(result) } else { None }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn slot(id: &str, product: &str, price: i64, qty: i64) -> Slot {
        Slot { slot_id: id.into(), product: product.into(),
               price_cents: price, quantity: qty }
    }

    fn setup() -> VendingMachine {
        let mut m = VendingMachine::new();
        m.add_slot(slot("A1", "Cola", 125, 5)).unwrap();
        m.add_slot(slot("C3", "Gum", 50, 10)).unwrap();
        m.load_change(25, 20).unwrap();
        m.load_change(10, 20).unwrap();
        m.load_change(5, 20).unwrap();
        m.load_change(1, 50).unwrap();
        m
    }

    #[test]
    fn select_refused_in_idle_state() {
        let mut m = setup();
        assert!(matches!(m.select("A1"), Err(VendingError::BadState(_))));
    }

    #[test]
    fn insert_coin_transitions_to_accepting() {
        let mut m = setup();
        assert_eq!(m.state, MachineState::Idle);
        m.insert_coin(25).unwrap();
        assert_eq!(m.state, MachineState::Accepting);
    }

    #[test]
    fn successful_purchase_dispenses_and_returns_change() {
        let mut m = setup();
        m.insert_coin(100).unwrap();
        m.insert_coin(25).unwrap();
        m.insert_coin(25).unwrap();
        let (product, change) = m.select("A1").unwrap();
        assert_eq!(product, "Cola");
        let change_sum: i64 = change.iter().map(|(d, c)| d * c).sum();
        assert_eq!(change_sum, 25);
        assert_eq!(m.state, MachineState::Idle);
        assert_eq!(m.inserted_cents, 0);
    }

    #[test]
    fn insufficient_funds_refused_but_state_preserved() {
        let mut m = setup();
        m.insert_coin(25).unwrap();
        let err = m.select("A1");
        assert!(matches!(err, Err(VendingError::InsufficientFunds { .. })));
        // State stays in Accepting — the user can still insert more or refund.
        assert_eq!(m.state, MachineState::Accepting);
        assert_eq!(m.inserted_cents, 25);
    }

    #[test]
    fn refund_returns_all_inserted_coins_and_resets() {
        let mut m = setup();
        m.insert_coin(100).unwrap();
        m.insert_coin(25).unwrap();
        let refunded = m.refund();
        let total: i64 = refunded.iter().map(|(d, c)| d * c).sum();
        assert_eq!(total, 125);
        assert_eq!(m.state, MachineState::Idle);
    }

    #[test]
    fn cannot_make_change_fails_cleanly() {
        let mut m = VendingMachine::new();
        m.add_slot(slot("G1", "Gum", 30, 10)).unwrap();
        // No small change loaded at all.
        m.insert_coin(100).unwrap();
        assert!(matches!(m.select("G1"), Err(VendingError::CannotMakeChange(_))));
    }
}
